tests/bofire/strategies/doe/test_utils.py

```diff
@@ -0,0 +1,664 @@
+            # opti.NonlinearInequality("x1**2 + x2**2 + x3**2 + x4**2 - 1"),
+            NChooseKConstraint(
+                features=["x1", "x2", "x3", "x4"],
+                max_count=3,
+                min_count=0,
+                none_also_valid=True,
+            ),
+        ],
+    )
+
+    x = np.array([[1, 1, 1, 1], [0.5, 0.5, 0.5, 0.5], [3, 2, 1, 0]]).flatten()
+
+    # linear equality
+    c = ConstraintWrapper(domain.constraints[0], domain, tol=0)
+    assert np.allclose(c(x), np.array([1.5, 0.5, 2.5]))
+
+    # linear inequaity
+    c = ConstraintWrapper(domain.constraints[1], domain, tol=0)
+    assert np.allclose(c(x), np.array([1.5, 0.5, 2.5]))
+
+    # # nonlinear equality
+    # c = ConstraintWrapper(domain.constraints[2], domain, tol=0)
+    # assert np.allclose(c(x), np.array([3, 0, 13]))
+
+    # # nonlinear inequality
+    # c = ConstraintWrapper(domain.constraints[3], domain, tol=0)
+    # assert np.allclose(c(x), np.array([3, 0, 13]))
+
+    # TODO: Not implemented error
+    # nchoosek constraint
+    c = ConstraintWrapper(domain.constraints[2], domain, tol=0)
+    assert np.allclose(c(x), np.array([1, 0.5, 0]))
+
+
+def test_d_optimality():
+    # define model matrix: full rank
+    X = np.array(
+        [
+            [1, 1, 0, 0],
+            [1, 0, 1, 0],
+            [1, 0, 0, 1],
+            [1, 0, 0, 0],
+        ]
+    )
+    assert np.allclose(d_optimality(X), np.linalg.slogdet(X.T @ X)[1])
+
+    # define model matrix: not full rank
+    X = np.array(
+        [
+            [1, 1, 0, 0],
+            [1, 0, 1, 0],
+            [1, 0, 0, 1],
+            [1, 1 / 3, 1 / 3, 1 / 3],
+        ]
+    )
+    assert np.allclose(d_optimality(X), np.sum(np.log(np.linalg.eigvalsh(X.T @ X)[1:])))
+
+
+def test_a_optimality():
+    # define model matrix: full rank
+    X = np.array(
+        [
+            [1, 1, 0, 0],
+            [1, 0, 1, 0],
+            [1, 0, 0, 1],
+            [1, 0, 0, 0],
+        ]
+    )
+    assert np.allclose(a_optimality(X), np.sum(1 / (np.linalg.eigvalsh(X.T @ X))))
+
+    # define model matrix: not full rank
+    X = np.array(
+        [
+            [1, 1, 0, 0],
+            [1, 0, 1, 0],
+            [1, 0, 0, 1],
+            [1, 1 / 3, 1 / 3, 1 / 3],
+        ]
+    )
+    assert np.allclose(a_optimality(X), np.sum(1 / (np.linalg.eigvalsh(X.T @ X)[1:])))
+
+
+def test_g_efficiency():
+    # define model matrix and domain: no constraints
+    X = np.array(
+        [
+            [1, 0, 0, 0],
+            [0, 0.1, 0, 0],
+            [0, 0, 0.1, 0],
+            [0, 0, 0, 0.1],
+        ]
+    )
+
+    domain = Domain(
+        input_features=[
+            ContinuousInput(key=f"x{i+1}", lower_bound=0.95, upper_bound=1.0)
+            for i in range(4)
+        ],
+        output_features=[ContinuousOutput(key="y")],
+    )
+    assert np.allclose(g_efficiency(X, domain), 0.333, atol=5e-3)
+
+    # define domain: sampling not implemented
+    domain = Domain(
```

Comment:
@jduerholt this would be an example where I do not know hat to do.

original question to Johannes: 
I am currently reimplementing tests from DoE to bofire. I noticed that in bofire NChooseKConstraints require to have inputs with lower_bounds==0. Additionally, LinearInequalityConstraints require two have at least two features – so there is no workaround. This blocks me from having NChooseKConstraints in all cases where inputs are not lower bounded by 0 - which seems restrictive to me...

Reply:
I understand you solution would be to add 
LinearEqualityConstraint(
                features=["x1", "x2", "x3", "x4"], coefficients=[-1, -1, -1, -1], rhs=-0.95
            ),
which would enforce the desired behavior as max_count=1 implies that only one variable can be active.

Reply:
Yep, it is implicit. But what does it mean for an NChooseK constraint if one of its substituents has a lower_bound>0? This is for me a bit ill defined, or not?

Reply:
Yes, I also discussed with @R-M-Lee  now and I see now what I want to do makes no sense. Still, I am pretty sure I have seen such ill defined problems in the old DoE package (e.g. I want chemical A or B, but if I have A i want to have at list X amount of it..). I think this makes sense if you add binary variables to the picture a_is_active, b_is_active and put NChooseK on those. But I don't think this needs to be solved by release. For now I suggest we drop support for such problems until we know how to solve then properly.

Reply:
Yep, good idea. I think this could be also a nice discussion for our next Hackathon ;)